DEVELOPMENT_BEST_PRACTICE.md

```diff
@@ -0,0 +1,308 @@
+
+Package level state is rarely required. If it is, it should be composed of re-usable constructs defined and exposed from the package itself.
+
+In avoiding package level state we avoid [hidden dependencies and unintended side effects](https://dave.cheney.net/practical-go/presentations/gophercon-israel.html#_avoid_package_level_state) stemming from global state mutation. Model loosely-coupled components by [declaring your dependencies](#dependencies).
+
+### Names
+
+##### Packages
+
+Package names help form the first impression of what a package provides and is used to reference the public API of the package. A good package name brings about clarity and purpose for consumers. Package names should be concise and represent the behavior the package provides not the types it contains. 
+
+Avoid stutter in struct and interface names by treating the package name as a namespace and considering how the types are consumed. For example, `drink.DrinkDecorator` should be `drink.Decorator`. You may find you have a type named after the package, soemtimes it's appropriate.
+
+If you find yourself using an appropriate package name that's commonly used as a variable, consider pluralising the package name.
+
+Avoid [relentlessly] `types`, `interfaces`, `common`, `util` or `base` packages. These names offer no indication on what the package provides beyond 'a bundle of stuff'. With time, these types of package turn into dumping grounds containing unrelated logic that can be the root cause for cyclic dependencies.  If a utility style package is required, make it specific. For example a `/utils/cmdutil` package may provide command utility functions.
+
+##### Function & method
+
+Functions and methods (functions with a receiver parameter) should adequetly describe their behavior. Generally, they should follow a verb-noun form. The types of parameters should help inform the function name.
+
+##### Variable
+
+Variable names should be concise and descriptive. Prefer single word names. The further away from the site of declaration a variable is used, the more descriptive it needs to be. For example, a variable named `n` used 30 lines after declaration makes it unnecessarily difficult to reason what it represents at the site of use.
+
+### <a name="dependencies"></a> Dependencies
+
+Functions should explicitly declare their dependencies as parameters. In doing so, they honor the [dependency inversion principle](https://www.digitalocean.com/community/conceptual_articles/s-o-l-i-d-the-first-five-principles-of-object-oriented-design#dependency-inversion-principle) and provide the necessary flexibility to write loosely coupled programs.
```

Comment:
This is totally correct but, reading it, gives me the impression that defining dependencies as arguments is enough to follow dependency inversion (even though you are not saying that explicitly).

Maybe this is just me (in that case ignore me), but I would maybe specify that doing this is necessary when following Dep Inversion but it's not enough.

Reply:
Its intentionally worded as 'honor'ing not satisfying for the reasons you purport. Perhaps there's a better word here that will clarify it, or perhaps the back half of the sentence is misleading.

Reply:
Totally, as I said, your wording is completely accurate. But I just wonder if I know this because I already knew the difference between one and the other.

One option is to add something like:

```suggestion
Functions should explicitly declare their dependencies as parameters. In doing so, they honor (but don't completely satisfy) the [dependency inversion principle](https://www.digitalocean.com/community/conceptual_articles/s-o-l-i-d-the-first-five-principles-of-object-oriented-design#dependency-inversion-principle) and provide the necessary flexibility to write loosely coupled programs.
```

Maybe this is just me, if no one else comments about this, feel free to ignore me.